File: backend/src/models/roulette/rouletteGame.model.js
const mongoose = require("mongoose");

const betSchema = new mongoose.Schema({
  betType:   { type: String, required: true },   // color | parity | half | dozen | column | number
  betValue:  { type: mongoose.Schema.Types.Mixed, required: true },
  amount:    { type: Number, required: true },
  won:       { type: Boolean, default: false },
  winAmount: { type: Number, default: 0 },
  payout:    { type: Number, default: 0 },       // e.g. 2, 5, 19
}, { _id: false });

const rouletteGameSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
    index: true,
  },
  username: {
    type: String,
    default: "Anonymous",
  },

  // ── Bets placed by player ─────────────────────────────────
  bets:     { type: [betSchema], required: true },
  totalBet: { type: Number, required: true },

  // ── Spin result ───────────────────────────────────────────
  spinResult: { type: Number, required: true },   // 0–36
  resultColor:  { type: String },                  // red | black | green
  resultParity: { type: String },                  // odd | even | zero
  resultHalf:   { type: String },                  // low | high | zero
  resultDozen:  { type: String },                  // 1st | 2nd | 3rd | zero
  resultColumn: { type: String },                  // col1 | col2 | col3 | zero

  // ── Outcome ───────────────────────────────────────────────
  totalWin:  { type: Number, default: 0 },
  result:    { type: String, enum: ["WIN", "LOSE"], required: true },
  winType:   { type: String, default: "normal" },   // normal | rare | popular | jackpot

  // ── Engagement ───────────────────────────────────────────
  freeSpinUsed: { type: Boolean, default: false },

  createdAt: { type: Date, default: Date.now },
});

module.exports = mongoose.model("RouletteGame", rouletteGameSchema);

File: frontend/lib/models/roulette/roulette_game_model.dart
// lib/models/roulette/roulette_game_model.dart

class RouletteBetResult {
  final String betType;
  final dynamic betValue;
  final int amount;
  final bool won;
  final int winAmount;
  final int payout;

  RouletteBetResult({
    required this.betType,
    required this.betValue,
    required this.amount,
    required this.won,
    required this.winAmount,
    required this.payout,
  });

  factory RouletteBetResult.fromJson(Map<String, dynamic> json) {
    return RouletteBetResult(
      betType:   json['betType']   ?? '',
      betValue:  json['betValue'],
      amount:    (json['amount']    as num?)?.toInt() ?? 0,
      won:       json['won']       == true,
      winAmount: (json['winAmount'] as num?)?.toInt() ?? 0,
      payout:    (json['payout']   as num?)?.toInt() ?? 0,
    );
  }
}

class RouletteGameModel {
  final String id;
  final String userId;
  final String username;
  final List<RouletteBetResult> bets;
  final int    totalBet;
  final int    spinResult;    // 0–36
  final String resultColor;   // red | black | green
  final String resultParity;  // odd | even | zero
  final String resultHalf;    // low | high | zero
  final String resultDozen;   // 1st | 2nd | 3rd | zero
  final String resultColumn;  // col1 | col2 | col3 | zero
  final int    totalWin;
  final String result;        // WIN | LOSE
  final String winType;       // normal | rare | popular | jackpot
  final bool   freeSpinUsed;
  final DateTime createdAt;

  RouletteGameModel({
    required this.id,
    required this.userId,
    required this.username,
    required this.bets,
    required this.totalBet,
    required this.spinResult,
    required this.resultColor,
    required this.resultParity,
    required this.resultHalf,
    required this.resultDozen,
    required this.resultColumn,
    required this.totalWin,
    required this.result,
    this.winType    = 'normal',
    this.freeSpinUsed = false,
    required this.createdAt,
  });

  bool get isWin => result == 'WIN';

  bool get isJackpot =>
      winType == 'jackpot' ||
      bets.any((b) => b.betType == 'number' && b.won);

  factory RouletteGameModel.fromJson(Map<String, dynamic> json) {
    return RouletteGameModel(
      id:           json['_id']          ?? '',
      userId:       json['userId']        ?? '',
      username:     json['username']      ?? 'Anonymous',
      bets:         (json['bets'] as List? ?? [])
                        .map((e) => RouletteBetResult.fromJson(e))
                        .toList(),
      totalBet:     (json['totalBet']     as num?)?.toInt() ?? 0,
      spinResult:   (json['spinResult']   as num?)?.toInt() ?? 0,
      resultColor:  json['resultColor']   ?? 'black',
      resultParity: json['resultParity']  ?? 'zero',
      resultHalf:   json['resultHalf']    ?? 'zero',
      resultDozen:  json['resultDozen']   ?? 'zero',
      resultColumn: json['resultColumn']  ?? 'zero',
      totalWin:     (json['totalWin']     as num?)?.toInt() ?? 0,
      result:       json['result']        ?? 'LOSE',
      winType:      json['winType']       ?? 'normal',
      freeSpinUsed: json['freeSpinUsed']  == true,
      createdAt:    json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}